core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateUnnest.java

```diff
@@ -0,0 +1,526 @@
+
+        @Override
+        public RewriteResult visitUnnest(UnnestNode node, Void context)
+        {
+            return new RewriteResult(sequenceSource, Optional.empty());
+        }
+
+        @Override
+        public RewriteResult visitEnforceSingleRow(EnforceSingleRowNode node, Void context)
+        {
+            // If EnforceSingleRowNode is present in the subquery, the type of rewritten UnnestNode is LEFT.
+            // Therefore, for each input row there is produced at least one row.
+            // To restore the semantics of EnforceSingleRowNode, it is sufficient to check that for each
+            // input row there is at most one row.
+            RewriteResult source = rewrite(node.getSource());
+
+            if (isScalar(source.getPlan(), lookup)) {
+                return source;
+            }
+
+            Symbol rowNumberSymbol;
+            PlanNode sourceNode;
+            if (source.getRowNumberSymbol().isPresent()) {
+                rowNumberSymbol = source.getRowNumberSymbol().get();
+                sourceNode = source.getPlan();
+            }
+            else {
+                rowNumberSymbol = symbolAllocator.newSymbol("row_number", BIGINT);
+                sourceNode = new RowNumberNode(
+                        idAllocator.getNextId(),
+                        source.getPlan(),
+                        ImmutableList.of(uniqueSymbol),
+                        false,
+                        rowNumberSymbol,
+                        Optional.of(2),
+                        Optional.empty());
+            }
+            ResolvedFunction fail = metadata.resolveFunction(QualifiedName.of("fail"), fromTypes(VARCHAR));
+            Expression predicate = new IfExpression(
+                    new ComparisonExpression(
+                            GREATER_THAN,
+                            rowNumberSymbol.toSymbolReference(),
+                            new GenericLiteral("BIGINT", "1")),
+                    new Cast(
+                            new FunctionCall(
+                                    fail.toQualifiedName(),
+                                    ImmutableList.of(new Cast(new StringLiteral("Scalar sub-query has returned multiple rows"), toSqlType(VARCHAR)))),
+                            toSqlType(BOOLEAN)),
+                    TRUE_LITERAL);
+
+            return new RewriteResult(new FilterNode(idAllocator.getNextId(), sourceNode, predicate), Optional.of(rowNumberSymbol));
+        }
+
+        @Override
+        public RewriteResult visitLimit(LimitNode node, Void context)
+        {
+            // To restore the semantics of LimitNode, subquery rows must be limited for each input row.
+            RewriteResult source = rewrite(node.getSource());
+
+            if (node.isWithTies()) {
+                return new RewriteResult(
+                        rewriteLimitWithTiesWithPartitioning(node, source.getPlan(), metadata, idAllocator, symbolAllocator, ImmutableList.of(uniqueSymbol)),
```

Comment:
~Can't we just leave `Limit` node here with replaced source and let `ImplementLimitWithTies` do the rest of the job?~
NVM :)